README.md
# Request ISNI

Python scripts to perform ISNI requests via a provided AtomPub API and report the outcome in CSV files.

ISNI is the [ISO 27729:2012](https://www.iso.org/standard/44292.html) standard name identifier that uniquely identifies public entities who contributed to creative works.
As an ISNI registration agency, the Royal Library of Belgium (KBR) can request ISNI identifiers from the ISNI central database.

The scripts of this repository are used to perform such requests against the ISNI [AtomPub API](https://isni.org/page/technical-documentation/): each request is one XML file containing the local KBR identifier of a public entity, the response XML files are stored, and a CSV report is generated showing the outcome of the performed requests, for example the newly assigned ISNI identifiers.

## Usage

Create and activate a Python virtual environment
```bash

# Create a new Python virtual environment
python3 -m venv py-request-isni-env

# Activate the virtual environment
source py-request-isni-env/bin/activate

# Install dependencies
pip -r requirements.txt
```

Given that you have correctly formatted ISNI request XML files in the folder `my-requests`, you can call the script in the following way to obtain response XML files in the folder `my-responses` and a report at `overview.csv`.

```bash
python request_isni.py \
    --url https://isni-m.oclc.org:/ATOM/isni?
    --input-folder my-requests \
    --response-folder my-responses \
    --output-file overview.csv
```

Please note, that you should be an ISNI member to use this API. If your IP is not whitelisted you will receive HTTP error messages.

Instead of providing the URL via commandline parameter, you can also create a `.env` file with the following content: `ISNI_ATOM_URL=https://isni-m.oclc.org:/ATOM/isni?`.

## Software tests

Functions in `lib.py` contain doctests, additionally there is an overal testfile `test_request_isni.py` to test the script `request_isni.py`.
The code coverage can be computed by executing the following commands (generated HTML files will highlight which paths in the code were not covered by the tests).

```bash 
# compute the code coverage from within the directory request_isni
coverage run test_request_isni.py`

# generate a human-friendly report in HTML
coverage html
```

## License

The license of this software was chosen using https://ufal.github.io/public-license-selector and based on licenses of software libraries used by this repo:

| Library | Description | License |
|---------|-------------|---------|
| certifi | Providing the list of Mozilla's carefully curated collection of Root certificates, such that we can communicate securely with the ISNI server via https. | [MPL 2.0](https://www.mozilla.org/en-US/MPL/2.0/) |
| charset-normalizer | A library to help reading text from an unknown charset encoding, this library is used by the `requests` library we are using. | [MIT](https://opensource.org/licenses/MIT) |
| coverage | A tool to compute the code coverage. We use this for quality control during our software tests, i.e. to ensure all paths in the code are covered during tests. | [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0) |
| idna | A library to handle internationalized domain names. It is used by `urllib3` and therefore indirectly by our project . | [BSD 3-clause](https://opensource.org/licenses/BSD-3-Clause) |
| python-dotenv | Functionality to load a .env environment file and make the environment variables accessible in Python. We use this library to provide the functionality of specifying the ISNI API URL in an environment variable instead of via a commandline parameter. | [BSD 3-clause](https://opensource.org/licenses/BSD-3-Clause) |
| requests | We use this library to perform HTTP requests against the ISNI APIs. | [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0) |
| urllib3 | Used by the requests library to make requests, thus implicitly used to make API requests against the ISNI APIs. | [MIT](https://opensource.org/licenses/MIT) |



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).


## [0.1.0] - 2022-12-20

### Added

- Script to request ISNI identifiers and keep a CSV log of the outcome

[0.1.0]: https://github.com/kbrbe/request-isni/releases/tag/v0.1.0


request_isni/test_request_isni.py
import importlib
import io
import sys
import shutil
import os
import csv
import unittest
import tempfile
import requests
import lib
import xml.etree.ElementTree as ET
from unittest.mock import patch, MagicMock
import request_isni

# Don't show the traceback of an AssertionError, because the AssertionError already says what the issue is!
__unittest = True

class TestRequestISNI(unittest.TestCase):

  # ---------------------------------------------------------------------------
  def setUp(self):
    self.tempOutput = os.path.join(tempfile.gettempdir(), 'output.csv')
    self.responseFolder = os.path.join(tempfile.gettempdir(), 'responses')

  # ---------------------------------------------------------------------------
  def tearDown(self):
    if os.path.isfile(self.tempOutput):
      os.remove(self.tempOutput)

    if os.path.exists(self.responseFolder):
      shutil.rmtree(self.responseFolder)
    

  # ---------------------------------------------------------------------------
  def testDoctests(self):
    """Test doctest unit tests of lib.py so it will be counted when computing the test coverage"""
    import doctest
    doctest.testmod(lib)

  # ---------------------------------------------------------------------------
  def testIncorrectInputFolder(self):
    """This function tests if the program stops if an invalid input folder is given."""
    with self.assertRaises(SystemExit):
      request_isni.main('folder-which-does-not-exist', 'outputFile.csv', 'folder-which-does-not-exist-responses', 'myURL')

  # ---------------------------------------------------------------------------
  @unittest.mock.patch('request_isni.requests.post')
  def testOutputGenerated(self, mock_request):
    """This function tests if a CSV with correct values is created based on a response from a mock API."""

    mock_response = MagicMock()
    mock_response.status_code = 200
    mock_response.text = '<responseRecord><ISNIAssigned><isniUnformatted>0000000000000001</isniUnformatted><ISNIMetadata><sources><codeOfSource>BNF</codeOfSource><sourceIdentifier>123</sourceIdentifier></sources></ISNIMetadata></ISNIAssigned></responseRecord>'
    mock_request.return_value = mock_response

    request_isni.main('testResources/oneXMLFile', self.tempOutput, self.responseFolder, 'myURL')

    with open(self.tempOutput, 'r') as outFile:
      resultReader = csv.reader(outFile)
      result = list(resultReader)

      expectedResult = [['KBRID', 'surname', 'forename', 'isniStatus', 'ISNI', 'PPN', 'SOURCE', 'comment'], ['123', 'Doe', 'John', 'confidentMatch', '0000 0000 0000 0001', '', '', '']]
      self.assertCountEqual(result, expectedResult, msg='Incorrect CSV output was produced')
      

  # ---------------------------------------------------------------------------
  @unittest.mock.patch('request_isni.requests.post')
  def testHTTPError(self, mock_request):
    """This function tests if an expected error messages is provided if the server response was not 200."""

    # preserving the original raise_for_status function by using the model of Response
    mock_response = requests.models.Response()

    # manually set the error values
    mock_response.status_code = 404
    mock_request.return_value = mock_response

    # capture print output from the script
    capturedOutput = io.StringIO()
    sys.stdout = capturedOutput

    # call the script
    request_isni.main('testResources/oneXMLFile', self.tempOutput, self.responseFolder, 'myURL')
    sys.stdout = sys.__stdout__

    expected = 'Found 1 files in directory from which 1 are XML files.\nProcess "file1.xml" ...\nThere was an HTTP response code which is not 200 for "file1.xml"\n404 Client Error: None for url: None\n\n'

    # check if what was printed was what was expected
    self.assertEqual(capturedOutput.getvalue(),
                     expected,
                     msg='Wrong error message')

if __name__ == '__main__':
  unittest.main()


request_isni/lib.py
#
# (c) 2022 Sven Lieber
# KBR Brussels
#
import xml.etree.ElementTree as ET

# -----------------------------------------------------------------------------
def formatISNI(isni):
  """This function formats ISNI identifiers by adding spaces if needed.

  >>> formatISNI("0001000100010001")
  '0001 0001 0001 0001'
  """
  isni = isni.strip()
  return isni[0:4] + ' ' + isni[4:8] + ' ' + isni[8:12] + ' ' + isni[12:16]

# -----------------------------------------------------------------------------
def parseResult(responseRecord):
  """This function parses an ISNI result record and extracts data.

  ------------------------------------------------------------
  # CASE CHAPTER 4: CONFIDENT MATCH
  When match is 100%

  If the ISNI matching found a confident match, we report the resulting ISNI identifier that we received.
  >>> xml1 = ET.fromstring('<responseRecord><ISNIAssigned><isniUnformatted>0000000000000001</isniUnformatted><ISNIMetadata><sources><codeOfSource>BNF</codeOfSource><sourceIdentifier>123</sourceIdentifier></sources></ISNIMetadata></ISNIAssigned></responseRecord>')
  >>> parseResult(xml1)
  {'isniStatus': 'confidentMatch', 'ISNI': '0000 0000 0000 0001', 'PPN': '', 'SOURCE': '', 'comment': ''}

  ------------------------------------------------------------
  # CASE CHAPTER 5: POSSIBLE MATCH (NoISNI)
  When match is betwen 060.00 and 085.00

  If the ISNI matching found a possible match, an ISNI-internal PPN identifier is returned which we add to the result
  >>> xml2 = ET.fromstring('<responseRecord><noISNI><possibleMatch><PPN>123</PPN><evaluationScore>70.00</evaluationScore><source>BNF#PERSON</source><mergeInstruction>P</mergeInstruction></possibleMatch></noISNI></responseRecord>')
  >>> parseResult(xml2)
  {'isniStatus': 'possibleMatch', 'ISNI': '', 'PPN': '123', 'SOURCE': 'BNF#PERSON', 'comment': ''}

  ------------------------------------------------------------
  # CASE CHAPTER 6: POSSIBLE MATCH CANNOT BE ASSIGNED (NoISNI)
  When a confident match is found, but the matching ISNI record is linked to another ISNI record

  >>> xml3 = ET.fromstring('<responseRecord><noISNI><PPN>123</PPN><reason>possible match cannot be assigned</reason><possibleMatch><PPN>456</PPN><evaluationScore>70.00</evaluationScore><source>BNF#PERSON</source><mergeInstruction>P</mergeInstruction></possibleMatch></noISNI></responseRecord>')
  >>> parseResult(xml3)
  {'isniStatus': 'possibleMatchCannotBeAssigned', 'ISNI': '', 'PPN': '456', 'SOURCE': 'BNF#PERSON', 'comment': ''}

  ------------------------------------------------------------
  # CASE CHAPTER 7: POSSIBLE MATCH (YOUR NAME IS ALREADY IN ISNI WITH ANOTHER ID)
  When a match was found, but the matching ISNI record already links to another of our local identifiers  

  For example, we sent a request to get an ISNI for KBRID 123, ISNI found a match, but that match already links to KBRID 456
  >>> xml4 = ET.fromstring('<responseRecord><noISNI><possibleMatch><PPN>456</PPN><evaluationScore>95.00</evaluationScore><source>KBR#PERSON</source><mergeInstruction>P</mergeInstruction></possibleMatch></noISNI></responseRecord>')
  >>> parseResult(xml4)
  {'isniStatus': 'possibleMatchDuplicateKBRID', 'ISNI': '', 'PPN': '456', 'SOURCE': 'KBR#PERSON', 'comment': ''}

  # CASE 26.2 Response NoISNI (Data not accepted / no reason given)
  >>> xml5 = ET.fromstring('<responseRecord><information>data not accepted</information><noISNI><reason>no match initial database</reason></noISNI></responseRecord>')
  >>> parseResult(xml5)
  {'isniStatus': 'noMatch', 'ISNI': '', 'PPN': '', 'SOURCE': '', 'comment': 'no match initial database'}

  # CASE 26.4 Response NoISNI (invalid data)
  >>> xml6 = ET.fromstring('<responseRecord><information>data not accepted</information><noISNI><reason>invalid data</reason><information>More info</information></noISNI></responseRecord>')
  >>> parseResult(xml6)
  {'isniStatus': 'noMatch', 'ISNI': '', 'PPN': '', 'SOURCE': '', 'comment': 'invalid data, More info'}
  
  """
  result = {'isniStatus': '', 'ISNI': '', 'PPN': '', 'SOURCE': '', 'comment': ''}
  if(responseRecord.tag == 'responseRecord'):
    
    isniAssigned = responseRecord.find('ISNIAssigned')

    result['comment'] = ''
    # There was a confident match and an ISNI identifier was returned
    if isniAssigned is not None:
      result['isniStatus'] = 'confidentMatch'
      result['ISNI'] = formatISNI(isniAssigned.find('isniUnformatted').text)
    else:

      # There should be noISNI indicating one of three cases
      isniNotAssigned = responseRecord.find('noISNI')
      if isniNotAssigned is not None:

        source = isniNotAssigned.find('possibleMatch/source')
        if source is not None:
          result['SOURCE'] = source.text
          # If we find a PPN as direct child there is a linked ISNI
          linkedISNIRecord = isniNotAssigned.find('PPN')
          if linkedISNIRecord is not None:
            result['isniStatus'] = 'possibleMatchCannotBeAssigned'
            result['PPN'] = isniNotAssigned.find('possibleMatch/PPN').text

          # It is a regular possible match or a possible match but with duplicate KBRID
          else:
            otherKBRRecord = True if isniNotAssigned.find('possibleMatch/source').text.startswith('KBR') else False
            if otherKBRRecord:
              result['isniStatus'] = 'possibleMatchDuplicateKBRID'
              result['PPN'] = isniNotAssigned.find('possibleMatch/PPN').text
            else:
              result['isniStatus'] = 'possibleMatch'
              result['PPN'] = isniNotAssigned.find('possibleMatch/PPN').text
 

        # There is no source given, probably no match found
        else:
          result['isniStatus'] = 'noMatch'
          reason = isniNotAssigned.find('reason')
          if reason is not None:
            result['comment'] = reason.text
          else:
            result['comment'] = 'No reason provided'

          # According to 26.4 there might be more info given if the reason is 'invalid data'
          # let's also add the additional info
          moreInfo = isniNotAssigned.find('information')
          if moreInfo is not None:
            result['comment'] = result['comment'] + ', ' + moreInfo.text

         
      # The record was no confident match and no 'NoISNI', what is it?
      else:
        print('This should not happen, unknown record structure:')
        print(ET.tostring(responseRecord, encoding='utf-8', method='xml'))
     
  return result

# -----------------------------------------------------------------------------
def getISNIRequestData(request):
  """This function extracts the KBRID from an ISNI request.

  >>> xml1 = ET.fromstring('<Request><identityInformation><requestorIdentifierOfIdentity><identifier>123</identifier></requestorIdentifierOfIdentity><identity><personOrFiction><personalName><surname>Doe</surname><forename>John</forename></personalName></personOrFiction></identity></identityInformation></Request>')
  >>> getISNIRequestData(xml1)
  {'KBRID': '123', 'surname': 'Doe', 'forename': 'John'}

  >>> xml2 = ET.fromstring('<myXML></myXML>')
  >>> getISNIRequestData(xml2)
  Traceback (most recent call last):
   ...
  Exception: Input is not an ISNI request
  """
  result = {'KBRID': '', 'surname': '', 'forename': ''}
  if request.tag == 'Request':
    result['KBRID'] = request.find('identityInformation/requestorIdentifierOfIdentity/identifier').text
    result['surname'] = request.find('identityInformation/identity/personOrFiction/personalName/surname').text
    result['forename'] = request.find('identityInformation/identity/personOrFiction/personalName/forename').text
  else:
    raise Exception('Input is not an ISNI request')

  return result
  


request_isni/request_isni.py
import requests
import csv
import os
from dotenv import load_dotenv
from datetime import date
import logging
import time
import xml.etree.ElementTree as ET
import urllib
import lib
from io import BytesIO
from optparse import OptionParser

NS_SRW = 'http://www.loc.gov/zing/srw/'


# -----------------------------------------------------------------------------
def checkArguments():

  parser = OptionParser(usage="usage: %prog [options]")
  parser.add_option('-i', '--input-folder', action='store', help='XML requests sent to the ISNI API are read from this folder')
  parser.add_option('-o', '--output-file', action='store', help='A list with results is stored in this CSV file')
  parser.add_option('-r', '--response-folder', action='store', help='The folder in which received responses are stored')
  parser.add_option('-u', '--url', action='store', help='The URL to which the request should be sent')
  (options, args) = parser.parse_args()

  #
  # Check if we got all required arguments
  #
  if( (not options.input_folder) or (not options.output_file) ):
    parser.print_help()
    exit(1)

  if not options.response_folder:
    options.response_folder = os.path.join(options.input_folder, 'responses')

  #
  # load environment variables from .env file
  #
  load_dotenv()

  if not options.url:
    options.url = os.getenv('ISNI_ATOM_URL')

  if options.url:
    return options
  else:
    print('No URL provided via commandline or environment file!')
    exit(1)

# -----------------------------------------------------------------------------
def main(inputFolder, outputFile, responseFolder, url):

  if not os.path.isdir(inputFolder):
    print(f'The provided inputFolder "{inputFolder}" is not a directory!')
    exit(1)

  # explicitly specify newline character empty to avoid problems on Windows-based systems
  with open(outputFile, 'w', encoding='utf-8', newline='') as outFile:
    outputWriter = csv.DictWriter(outFile, fieldnames=['KBRID', 'surname', 'forename', 'isniStatus', 'ISNI', 'PPN', 'SOURCE', 'comment'])
    outputWriter.writeheader()

    if not os.path.exists(responseFolder):
      os.makedirs(responseFolder)

    #
    # collect all relevant ISNI requests files from the input directory
    #
    allFiles = os.listdir(inputFolder)
    allXMLFiles = []
    for inputFile in os.listdir(inputFolder):
      if inputFile.endswith('.xml'):
        allXMLFiles.append(inputFile)

    numberFiles = len(allFiles)
    numberXMLFiles = len(allXMLFiles)
    print(f'Found {numberFiles} files in directory from which {numberXMLFiles} are XML files.')

    for inputFile in allXMLFiles:

      inputData = None
      responseBytes = None
      responseText = None

      #
      # First step: check the request and extract data we need for the output: the KBRID
      #
      with open(os.path.join(inputFolder, inputFile), 'rb') as inFile:

        try:
          print(f'Process "{inputFile}" ...')

          inputData = lib.getISNIRequestData(ET.parse(inFile).getroot())
        except Exception as e:
          print(f'Error reading/parsing the input file "{inputFile}"')
          print(e)
          continue

      #
      # Second step: send the request to ISNI and parse the response
      #
      with open(os.path.join(inputFolder, inputFile), 'rb') as inFile, \
           open(os.path.join(responseFolder, f'response-{inputFile}'), 'w', encoding='utf-8') as responseFile:

        response = None
        try:
          # try to send one ISNI XML request to the server
          response = requests.post(url, data=inFile)

          # when everything was okay, save the response in case we have to lookup details later
          responseBytes = response.content

          if response.encoding is None:
            response.encoding = 'utf-8'
          responseText = response.text
          responseFile.write(responseText)

          # When the API did not return HTTP code 200 OK, but for example the error code 404 or 500
          # then raise an error (so we do not have to manually check if an error occurred,
          # instead we can handle the raised error using the except blocks below)
          response.raise_for_status()

        except requests.exceptions.Timeout:
          print(f'There was a timeout when sending "{inputFile}"')
        except requests.exceptions.TooManyRedirects:
          print(f'There were too many redirects for "{inputFile}"')
          continue
        except requests.exceptions.HTTPError as err:
          print(f'There was an HTTP response code which is not 200 for "{inputFile}"')
          print(err)
          print(responseText)
          continue
        except UnicodeEncodeError as err:
          print(err)
          print(f'There was an issue with the encoding of the ISNI server response, it should have been "{response.encoding}"')
          rawResponseFilename = os.path.join(responseFolder, f'response-raw-{inputFile}')
          with open(rawResponseFilename, 'wb') as responseBytes:
            responseBytes.write(response.content)
          continue
        except requests.exceptions.RequestException as err:  
          print(f'Something went wrong when trying to send a request for "{inputFile}"')
          print(err)
          continue

      #
      # Third step: combine data from the request and the response to append a row to our output CSV
      #
      try:
        # when everything was okay, parse the resulting XML and add information to the output file
        outputData = lib.parseResult(ET.fromstring(responseText))

        # write a combination of input data (the KBRID for which an ISNI was requested)
        # and output data (fields from the ISNI request response) to the output file
        outputData.update(inputData)
        outputWriter.writerow(outputData)
      except Exception as err:
        print(f'There was an error when adding a line to the output file "{outputFile}" based on request "{inputFile}"')
        print(err)


     
if __name__ == '__main__':
  args = checkArguments()
  main(args.input_folder, args.output_file, args.response_folder, args.url)
